blog/_src/posts/2017-11-XX-adding-refinement-types.md

```diff
@@ -0,0 +1,449 @@
+the range specified by `start` and `end` is some sensible (although
+possibly empty) portion of the input vector `vec`. A `Refine` allows
+us to indicate that the length of the returned vector will always be
+`(- end start)`:
+
+
+```racket
+(: vector-copy
+   (All (A) (-> ([vec : (Vectorof A)]
+                 [start : Natural]
+                 [end : Natural])
+                #:pre (vec start end)
+                (<= start end (vector-length vec))
+                (Refine [res : (Vectorof A)]
+                        (= (- end start)
+                           (vector-length res))))))
+(define (vector-copy vec start end)
+  (define len (- end start))
+  (cond
+    [(= 0 len) (vector)]
+    [else
+     (define res (make-vector len (safe-vector-ref vec start)))
+     (let loop! ([i : Natural 0])
+       (when (< i len)
+         (define a (safe-vector-ref vec (+ start i)))
+         (safe-vector-set! res i a)
+         (loop! (+ 1 i))))
+     res]))
+```
+
+`vector-copy!` is similar, but instead of returning a fresh vector,
+the elements in the specified range are copied from the vector `src`
+into the vector `dest`:
+
+
+```racket
+> (define src (vector "p" "e" "a" "r"))
+> (define dest (vector "" "" "" "" ""))
+> (vector-copy! dest 1 src 1 4)
+> dest
+'#("" "e" "a" "r" "")
+```
+
+Here the precondition is larger because it includes constraints that
+must hold with respect to both vectors `src` and `dest`, i.e. the
+range in the `src` vector [`src-start`,`src-end`) must be a valid
+range of elements, and the destination vector `dest` must have a
+corresponding valid range [`dest-start`,`(+ dest-start (- src-end
+src-start))`):
+
+
+```racket
+(: vector-copy!
+   (All (A) (-> ([dest : (Vectorof A)]
+                 [dest-start : Natural]
+                 [src : (Vectorof A)]
+                 [src-start : Natural]
+                 [src-end : Natural])
+                #:pre (dest dest-start src src-start src-end)
+                (and (<= src-start src-end (vector-length src))
+                     (<= dest-start (vector-length dest))
+                     (<= (- src-end src-start)
+                         (- (vector-length dest) dest-start)))
+                Void)))
+(define (vector-copy! dest dest-start src src-start src-end)
+  (define count (- src-end src-start))
+  (let loop! ([i : Natural 0])
+    (when (< i count)
+      (define a (safe-vector-ref src (+ src-start i)))
+      (safe-vector-set! dest (+ dest-start i) a)
+      (loop! (+ 1 i)))))
+```
+
+## Typed/Untyped interaction
+
+Typed Racket will generate dependent function contracts (e.g. `->i`
+contracts) to protect dependent functions when they are used in
+untyped modules:
+
+```racket
+#lang racket/base
+
+(require "safe-vector-ops.rkt")
+
+(safe-vector-ref (vector 0 1) 2)
+
+;; produces:
+;;
+;; safe-vector-ref: contract violation
+;;  #:pre condition violation; ...
+```
+
+As always, care should be taken when designing APIs that involve
```

Comment:
suggestion: replace this paragraph with something like "the contract will check the arguments at runtime".

(Better to say _why_ care should be taken than to just say "beware")